pkg/scalers/gcp_storage_scaler.go

```diff
@@ -0,0 +1,230 @@
+	maxBucketItemsToScan int
+	metricName           string
+	targetObjectCount    int
+	scalerIndex          int
+}
+
+var gcsLog = logf.Log.WithName("gcp_storage_scaler")
+
+// NewGcsScaler creates a new gcsScaler
+func NewGcsScaler(config *ScalerConfig) (Scaler, error) {
+	meta, err := parseGcsMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing GCP storage metadata: %s", err)
+	}
+
+	ctx := context.Background()
+
+	var client *storage.Client
+
+	switch {
+	case meta.endpoint != "":
+		client, err = storage.NewClient(ctx, option.WithEndpoint(meta.endpoint), option.WithoutAuthentication())
+	case meta.gcpAuthorization.podIdentityProviderEnabled:
+		client, err = storage.NewClient(ctx)
+	case meta.gcpAuthorization.GoogleApplicationCredentialsFile != "":
+		client, err = storage.NewClient(
+			ctx, option.WithCredentialsFile(meta.gcpAuthorization.GoogleApplicationCredentialsFile))
+	default:
+		client, err = storage.NewClient(
+			ctx, option.WithCredentialsJSON([]byte(meta.gcpAuthorization.GoogleApplicationCredentials)))
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("storage.NewClient: %v", err)
+	}
+
+	bucket := client.Bucket(meta.bucketName)
+	if bucket == nil {
+		return nil, fmt.Errorf("failed to create a handle to bucket %s", meta.bucketName)
+	}
+
+	gcsLog.Info(fmt.Sprintf("Metadata %v", meta))
+
+	return &gcsScaler{
+		client:   client,
+		bucket:   bucket,
+		metadata: meta,
+	}, nil
+}
+
+func parseGcsMetadata(config *ScalerConfig) (*gcsMetadata, error) {
+	meta := gcsMetadata{}
+	meta.targetObjectCount = defaultTargetObjectCount
+	meta.maxBucketItemsToScan = defaultMaxBucketItemsToScan
+
+	if val, ok := config.TriggerMetadata["bucketName"]; ok {
+		if val == "" {
+			gcsLog.Error(nil, "no bucket name given")
+			return nil, fmt.Errorf("no bucket name given")
+		}
+
+		meta.bucketName = val
+	} else {
+		gcsLog.Error(nil, "no bucket name given")
+		return nil, fmt.Errorf("no bucket name given")
+	}
+
+	if val, ok := config.TriggerMetadata["targetObjectCount"]; ok {
+		targetObjectCount, err := strconv.Atoi(val)
+		if err != nil {
+			gcsLog.Error(err, "Error parsing targetObjectCount")
+			return nil, fmt.Errorf("error parsing targetObjectCount: %s", err.Error())
+		}
+
+		meta.targetObjectCount = targetObjectCount
+	}
+
+	if val, ok := config.TriggerMetadata["maxBucketItemsToScan"]; ok {
+		maxBucketItemsToScan, err := strconv.Atoi(val)
+		if err != nil {
+			gcsLog.Error(err, "Error parsing maxBucketItemsToScan")
+			return nil, fmt.Errorf("error parsing maxBucketItemsToScan: %s", err.Error())
+		}
+
+		meta.maxBucketItemsToScan = maxBucketItemsToScan
+	}
+
+	if val, ok := config.TriggerMetadata["endpoint"]; ok {
+		if val != "" {
+			meta.endpoint = val
+		}
+	}
+
+	if meta.endpoint == "" {
+		auth, err := getGcpAuthorization(config, config.ResolvedEnv)
+		if err != nil {
+			return nil, err
+		}
+		meta.gcpAuthorization = auth
+	}
+
+	if val, ok := config.TriggerMetadata["metricName"]; ok {
```

Comment:
We shouldn't use `metricName` from the trigger config, we should just generate the metric name.

Reply:
I saw the option to specify an alternate metric name in other scalers as [well](https://github.com/kedacore/keda/blob/b6bc7b754e2b5145c45895126e9dc57e86ab8816/pkg/scalers/azure_blob_scaler.go#L120)

Reply:
That's right, but we are trying to get rid of this (as it is just an internal implementation detail) and now we have a secure way to generate unique metric name.